src/callback_handler/kubernetes/client.rs

```diff
@@ -0,0 +1,241 @@
+use anyhow::{anyhow, Result};
+use kube::core::{DynamicObject, ObjectList};
+use std::collections::HashMap;
+use tokio::sync::RwLock;
+
+use crate::callback_handler::kubernetes::{reflector::Reflector, ApiVersionKind, KubeResource};
+
+pub(crate) struct Client {
+    kube_client: kube::Client,
+    kube_resources: RwLock<HashMap<ApiVersionKind, KubeResource>>,
+    reflectors: RwLock<HashMap<String, Reflector>>,
+}
+
+impl Client {
+    pub fn new(client: kube::Client) -> Self {
+        Self {
+            kube_client: client,
+            kube_resources: RwLock::new(HashMap::new()),
+            reflectors: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Build a KubeResource using the apiVersion and Kind "coordinates" provided.
+    /// The result is then cached locally to avoid further interactions with
+    /// the Kubernetes API Server
+    async fn build_kube_resource(&mut self, api_version: &str, kind: &str) -> Result<KubeResource> {
+        let avk = ApiVersionKind {
+            api_version: api_version.to_string(),
+            kind: kind.to_string(),
+        };
+
+        // take a reader lock and search for the resource inside of the
+        // known resources
+        let kube_resource = {
+            let known_resources = self.kube_resources.read().await;
+            known_resources.get(&avk).map(|r| r.to_owned())
+        };
+        if let Some(kr) = kube_resource {
+            return Ok(kr);
+        }
+
+        // the resource is not known yet, we have to search it
+        let resources_list = match api_version {
+            "v1" => {
+                self.kube_client
+                    .list_core_api_resources(api_version)
+                    .await?
+            }
+            _ => self
+                .kube_client
+                .list_api_group_resources(api_version)
+                .await
+                .map_err(|e| anyhow!("error finding resource {api_version} / {kind}: {e}"))?,
+        };
+
+        let resource = resources_list
+            .resources
+            .iter()
+            .find(|r| r.kind == kind)
+            .ok_or_else(|| anyhow!("Cannot find resource {api_version}/{kind}"))?
+            .to_owned();
+
+        let (group, version) = match api_version {
+            "v1" => ("", "v1"),
+            _ => api_version
+                .split_once('/')
+                .ok_or_else(|| anyhow!("cannot determine group and version for {api_version}"))?,
+        };
+
+        let kube_resource = KubeResource {
+            resource: kube::api::ApiResource {
+                group: group.to_string(),
+                version: version.to_string(),
+                api_version: api_version.to_string(),
+                kind: kind.to_string(),
```

Comment:
```suggestion
                group: group.to_owned(),
                version: version.to_owned(),
                api_version: api_version.to_owned(),
                kind: kind.to_owned(),
```



[Use `to_owned()`. ](https://github.com/kubewarden/.github/blob/main/CONTRIBUTING.md#rust-code-conventions)